ORT fuzz testing

satyajandhyala opened the pull request:
**Description**: Add support to run fuzz testing of onnxruntime ORT flow (loading a model and running inference). The fuzz testing code should mutate a ORT (flatbuffer) model, exercise loading the model and inference, and capture any exceptions raised by onnxruntime.  Fix any unexpected/uncaught exception that would cause onnxruntime fail unexpectedly. For example, if the input file is missing or incorrect then terminating inference is expected; whereas indexing out-of-range of an STL collection class resulting either abrupt termination or unexpected behavior should be exposed by the fuzz testing and the onnxruntime code should be fixed.

**Motivation and Context**: The onnxruntime may be used on the corporate servers running critical applications. The idea to address any threats to the server by way of running inference on a badly modified ORT model to make onnxruntime crash and somehow use the crash to compromise the serve. This intent of this code to stress test against such attack and identify and fix an vulnerabilities. 
- Why is this change required? What problem does it solve?
This change is a test that should run as a build pipeline so that any code change in future would be tested using the scenarios described above.
- If it fixes an open issue, please link to the issue here.
https://msdata.visualstudio.com/Vienna/_workitems/edit/916254

gwang-msft commented on onnxruntime/core/session/inference_session.cc:
```diff
@@ -1151,9 +1151,17 @@ common::Status InferenceSession::Initialize() {
+    // option and ignore the serialized session state.
+    const experimental::fbs::SessionState* serialized_session_state = nullptr;
+#if defined(ORT_MINIMAL_BUILD)
     auto* serialized_session_state = !ort_format_model_bytes_.empty()
```
> auto* [](start = 3, length = 6)

Remove?

satyajandhyala replied:
I see the problem(multiple definitions) when compiling with ORT_MINIMAL_BUILD.

---
In reply to: [521617571](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r521617571) [](ancestors = 521617571)

gwang-msft commented on onnxruntime/test/fuzzing/src/OnnxPrediction.cpp:
```diff
@@ -2,40 +2,57 @@
+}
+
+OnnxPrediction::OnnxPrediction(const std::vector<char>& modelData)
+    : numBytes(modelData.size()),
```
>numBytes [](start = 6, length = 8)

Seems this is only used in the constructor, do you need this to be a member variable?

satyajandhyala replied:
That is correct. I had some other changes which I didn't commit, that required the numBytes as a member. I will remove from this commit.

skottmckay commented on onnxruntime/test/framework/ort_model_only_test.cc:
```diff
@@ -237,6 +237,33 @@ static void DumpOrtModelAsJson(const std::string& model_uri) {
 }
 */
 
+TEST(OrtModelOnlyTests, SerializeMnistToOrtFormat) {
```
What's the value of this test? Is it there some functionality it is covering that the existing tests don't?

satyajandhyala replied:
I ran into errors while running fuzz test using the MNIST example where the other testcase ran okay. You have provided me the fix and then MNIST worked fine. Because this example uncovered a problem. I thought that it is a good idea to keep the Mnist test.

---
In reply to: [521676385](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r521676385) [](ancestors = 521676385)

skottmckay replied:
If you remove the change to inference_session.cc does this test fail? 

IIRC that required the ORT model being created with an optimization level of 2, and that model being loaded in a full build where we perform level 3 optimizations which change some nodes, but attempt to use the kernel hashes from the ORT model. I didn't notice the necessary steps here to replicate that scenario though.

The test should be commented to explain exactly what is being tested if it's for that edge case. Otherwise the test is likely to be removed as it's unclear what value is being added. I'd suggest the test name shouldn't be about 'mnist' but instead about loading an ORT format model in a full build where additional optimizations may be applied. 

Does LoadMnistOrtFormatModel test add value? The 'load only' tests are all we can run in a minimal build, so no optimizers would run in that case.

---
In reply to: [521713904](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r521713904) [](ancestors = 521713904,521676385)

satyajandhyala replied:
Yes. It did fail when I removed the change to inference_sesion.cc

---
In reply to: [521808460](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r521808460) [](ancestors = 521808460,521713904,521676385)

satyajandhyala replied:
I think it is better make the flow independent of optimization settings. With your fix in inference_session.cc it is working now without choosing optimization level 2. This test case is to cover the scenario as the other model, ort_github_issue_4031.onnx.ort, did not catch the issue.

---
In reply to: [521808460](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r521808460) [](ancestors = 521808460,521713904,521676385)

skottmckay replied:
ValidateOrtFormatModelDoesNotRunOptimizersInFullBuild

---
In reply to: [522446895](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r522446895) [](ancestors = 522446895,521808460,521713904,521676385)

skottmckay commented:
Is changing the onnx commit id intentional?

skottmckay commented on cmake/onnxruntime_fuzz_test.cmake:
```diff
@@ -40,6 +40,7 @@ if (onnxruntime_FUZZ_ENABLED)
 	
 	# compile with c++17
 	target_compile_features(onnxruntime_security_fuzz PUBLIC cxx_std_17)
+	target_compile_options(onnxruntime_security_fuzz PRIVATE "/wd5208")
```
Add comment saying why we are disabling this warning

skottmckay commented on onnxruntime/core/session/inference_session.cc:
```diff
@@ -1151,9 +1151,17 @@ common::Status InferenceSession::Initialize() {
+    // graph, which would result in it no longer matching the saved session state.
+    // as it's a full build we can create the SessionState as per usual using kernel lookups, so we can take the safe
+    // option and ignore the serialized session state.
+    const experimental::fbs::SessionState* serialized_session_state = nullptr;
```
We're going to go with a different approach to handling ORT format models (PR that is part of should go out today). Is this change required for the PR to work or can it be reverted for now?

satyajandhyala replied:
We need this declaration for the regular build (non minimal build) to compile. I will remove the "auto*" in the following #ifdef block as Guoyu suggest.

---
In reply to: [521677655](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r521677655) [](ancestors = 521677655)

skottmckay commented:
What's the reason for putting the .ort file in a different place to all the other test data? Would have expected it to be in onnxruntime/test/testdata/mnist.ort so it's with mnist.onnx

skottmckay commented on onnxruntime/test/fuzzing/include/OnnxPrediction.h:
```diff
@@ -46,17 +46,20 @@ class OnnxPrediction
+  // Memory buffer pointing to the model
+  //
+  OnnxPrediction(const std::vector<char>& modelData);
+
```
nit: model_data

A lot of the variable and function naming in this file does not conform to the coding standards, so perhaps that can be fixed in a separate PR.

satyajandhyala replied:
Are we following the camel naming convention for the variables. In this file I found both camel convention and "_" separate variables. I should read the coding style document. Is the on the group wiki?

satyajandhyala replied:
fixed the variable anmes

skottmckay commented on onnxruntime/test/fuzzing/include/OnnxPrediction.h:
```diff
@@ -46,17 +46,20 @@ class OnnxPrediction
+
+  // Deletes the prediction object
   //
   ~OnnxPrediction();
```
nit: unnecessary comment. anybody working on C++ should know what a destructor does.

skottmckay commented on onnxruntime/test/fuzzing/src/OnnxPrediction.cpp:
```diff
@@ -45,101 +62,83 @@ session{nullptr}
 
```
It's confusing to be using the one OnnxPrediction class for both ONNX and ORT format models. Can we lift the common functionality into a Prediction class, and derive or compose (prefer compose if possible) an OnnxPrediction and OrtFormatPrediction class from that? An 'OnnxPrediction' ctor that takes a buffer and internally forces that to be treated as ORT could be unexpected.

satyajandhyala replied:
The class has two constructors one used by the ONNX flow (ModelProto) and the other use by ORT flow (flatbuffer passed asvector<char>). Other than that we are not doing any thing specific. The constructors create the Session object in both cases using the model and the other functionality works with the session object. I am not sure in this case we require a class hierarchy. We could take this up later if we add code specific to ORT or ONNX.

skottmckay commented on onnxruntime/test/fuzzing/src/OnnxPrediction.cpp:
```diff
@@ -45,101 +62,83 @@ session{nullptr}
+    //
+    ptrSession.release();
   }
 }
```
This seems somewhat egregious. I know it was existing code, but can we clean up the ownership semantics here? 

Whoever owns the session should have it in a unique_ptr and let that free it. Whoever is using the session but does not own it should just have a reference to the session. I don't really know what 'deleting the stack' means either. :)

satyajandhyala replied:
The session_ptr is nullptr if the OnnxPrediction object is created from the model file rather than the ModelProto. It seems the code is making sure that ptr_session actually points to the session object. Either session_ptr is nullptr or there is a session object created and needs to be "destroyed". In the OnnxPrediction::~OnnxPrediction function we can safely check and delete (reset) ptr_session.

---
In reply to: [521682737](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r521682737) [](ancestors = 521682737)

skottmckay commented on onnxruntime/test/fuzzing/src/OnnxPrediction.cpp:
```diff
@@ -2,40 +2,57 @@
+                                     this->GetAllocator().Free(ptr);
+                                   }};
+  memcpy(rawModel.get(), modelData.data(), numBytes);
+  Ort::SessionOptions so;
```
Do we need this complication? I assume the test runner will create a buffer that will remain valid whilst this instance is in use, or it can transfer ownership of that buffer here via a unique_ptr. Based on that I don't think we need to do another copy here, and even if we did I'm not aware of a reason why we'd need to complicate the allocation by using an ORT allocator.

skottmckay commented on onnxruntime/test/fuzzing/src/OnnxPrediction.cpp:
```diff
@@ -2,40 +2,57 @@
+    : session{nullptr} {
+  size_t numBytes = modelData.size();
+  rawModel = std::shared_ptr<void>{alloc.Alloc(numBytes),
+                                   [this](void* ptr) {
```
shared_ptr should rarely be needed. avoid wherever possible.

satyajandhyala replied:
Due to the custom destructor, I could not make it a unique_ptr. We want to have this memory managed i guess instead of freeing memory in the the destructor. Let me know if we want to make raw_model a raw pointer. 

---
In reply to: [521684570](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r521684570) [](ancestors = 521684570)

skottmckay commented:
> 

Seems odd to 'resize' (vs. 'reserve') and to not populate these if we're going to use them. 

---
Refers to: onnxruntime/test/fuzzing/src/OnnxPrediction.cpp:59 in 221eebd. [](commit_id = 221eebdae305cc046020727d4649b44d6f556cc1, deletion_comment = False)

skottmckay commented on onnxruntime/test/fuzzing/src/test.cpp:
```diff
@@ -180,61 +165,52 @@ int processCommandLine(int argc, char* argv[], runtimeOpt& opt)
           opt.userOpt.stress = true;
+        } else if (option == "/f") {
+          opt.userOpt.flatbuffer = true;
         }
```
would be better to name based on what it is (ort format model) vs. the current internal storage mechanism (flatbuffer)

skottmckay commented on onnxruntime/test/fuzzing/src/test.cpp:
```diff
@@ -243,39 +219,58 @@ int processCommandLine(int argc, char* argv[], runtimeOpt& opt)
+static size_t num_unknown_exception{0};
+static size_t num_successful_runs{0};
+static size_t iteration{0};
+
```
Given this is C++ can we wrap this in a class or struct instead of using file scope statics? You can have an instance of this class in main() so it's accessible in the 'catch' statement that tracks these.

skottmckay commented on onnxruntime/test/fuzzing/src/test.cpp:
```diff
@@ -243,39 +219,58 @@ int processCommandLine(int argc, char* argv[], runtimeOpt& opt)
+static void fuzz_handle_exception() {
+  try {
+    throw;
+  } catch (const Ort::Exception& ortException) {
```
Do we need this as a separate function with a re-throw? Is it re-used somewhere?

skottmckay commented on onnxruntime/test/fuzzing/src/test.cpp:
```diff
@@ -285,123 +280,113 @@ int main(int argc, char* argv[])
+      }
+    } else {
+      std::wstring ortModelFile = modelFile + L".ort";
+      Ort::SessionOptions so;
```
Shouldn't we just use the specified model filename as is?

satyajandhyala replied:
The model_file is the input ONNX file. We are making sure that we run this test using the ORT file generated from the ONNX file so that we use the latest changes, instead of a save ORT checked in to the repo. The ort_model file name generated by adding ".ort" to the input ONNX filename.

---
In reply to: [521690308](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r521690308) [](ancestors = 521690308)

skottmckay replied:
This code does not generate the .ort file does it? If this code is just testing the input it is provided, why is it imposing some arbitrary rule about file naming that isn't visible to a user (i.e. the input should be an ort file that must have an .ort extension, but that extension should not be passed when running the program). It may make sense if we were always testing a pair of files, but afaik each test is a standalone test of a single input, be that an ONNX or an ORT format model.

satyajandhyala replied:
Because of the option kOrtSessionOptionsConfigSaveModelFormat and SetOptimizedModelPath the first session dump the ORT model to the specified file. The ortModelStream reads the dumped ORT file into model_data. We want to make sure sure that we are also using the latest changes/fixes in onnxruntime to save the or ORT model from ONNX rather than an outdate ORT file.

---
In reply to: [522423493](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r522423493) [](ancestors = 522423493)

skottmckay commented on onnxruntime/test/fuzzing/src/test.cpp:
```diff
@@ -285,123 +280,113 @@ int main(int argc, char* argv[])
+        char tmp = modelData[i];
+        int j = (numBytes - i - 1) ? (i + 1) : 0;  // j = (i+1) % (numBytes -1)
+        modelData[i] ^= modelData[j];
+        modelData[i] = tmp;
```
Comment explaining what we're doing and why would be good. We may want to expand the types of tests in the future as well instead of just xor of 2 bytes (e.g. swap 8 bytes between 2 locations).

skottmckay commented on onnxruntime/test/fuzzing/src/test.cpp:
```diff
@@ -285,123 +280,113 @@ int main(int argc, char* argv[])
+          predict.SetupInput(GenerateDataForInputTypeTensor, 0);
+          predict.RunInference();
+          predict.PrintOutputValues();
+          num_successful_runs++;
```
What's the reason for printing output values?

satyajandhyala replied:
My understanding is, the test runs through all the mutations capturing all the exceptions whether or not handled by onnxruntime and finally produces a the summary, which may inspected to know whether or not all mutations generated any errors.

---
In reply to: [521692691](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r521692691) [](ancestors = 521692691)

satyajandhyala replied:
This was in the original code. Yes it dumps large tensors to the stdout. Unless it is used, I will go ahead and remove it.

skottmckay commented on onnxruntime/test/fuzzing/src/test.cpp:
```diff
@@ -285,123 +280,113 @@ int main(int argc, char* argv[])
+      ortModelStream.read(modelData.data(), numBytes);
+      ortModelStream.close();
+      for (int i = 8; i < numBytes; i++) {
+        char tmp = modelData[i];
```
What's the reason for skipping the first 8 bytes?

satyajandhyala replied:
Originally I had run into errors due to not finding "ORTM" and a magic number at the beginning. I thought I should skip the fist bytes. Changing it back to iterate from 1.

---
In reply to: [521693301](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r521693301) [](ancestors = 521693301)

skottmckay commented on onnxruntime/test/fuzzing/src/test.cpp:
```diff
@@ -285,123 +280,113 @@ int main(int argc, char* argv[])
+      for (int i = 8; i < numBytes; i++) {
+        char tmp = modelData[i];
+        int j = (numBytes - i - 1) ? (i + 1) : 0;  // j = (i+1) % (numBytes -1)
+        modelData[i] ^= modelData[j];
```
can we just start i at 1 in the 'for' loop and use j = i - 1?

skottmckay commented on onnxruntime/test/fuzzing/src/test.cpp:
```diff
@@ -285,123 +280,113 @@ int main(int argc, char* argv[])
+        int j = (numBytes - i - 1) ? (i + 1) : 0;  // j = (i+1) % (numBytes -1)
+        modelData[i] ^= modelData[j];
+        modelData[i] = tmp;
+        try {
```
Doesn't this mean you undo the mutation before you attempt to run the model?

skottmckay replied:
This refers to `modelData[i] = tmp;`.

skottmckay commented on onnxruntime/test/fuzzing/src/test.cpp:
```diff
@@ -285,123 +280,113 @@ int main(int argc, char* argv[])
-  catch(...)
-  {
-    Logger::testLog<< L"Something Went very wrong: " << Logger::endl;
+  } catch (const Ort::Exception& ortException) {
```
If there's a non-ORT error should this return a non-zero value?

skottmckay replied:
Sorry - the diffs make it a bit hard to read without using an external diff too. Looks like it's similar to the old logic, but not quite following why we have a catch for an OrtException outside of the test loop. What is actually being tested and expected is unclear. I would have expected there was a section of code with all the ORT calls that is tested. Within that section an OrtException is fine, a std::exception is probably ok but not expected, and anything else is bad. In the main program, I would have thought any exception was bad and coming from the early exit path in the test code where it rethrows in the catch-all.

---
In reply to: [521695691](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r521695691) [](ancestors = 521695691)

satyajandhyala replied:
Broadly speaking the try-catch block in inner loop is to catch the onnxruntime related exceptions and outer try-catch block is catch exceptions such as file-not-found, permissions, etc.

---
In reply to: [521718084](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r521718084) [](ancestors = 521718084,521695691)

satyajandhyala commented:
I only added mnist.onnx model under onnxruntime/test/testdata. The ORT file get generated at run time from the ONNX file under build\Windows\Debug\Debug\testdata by appending the suffix (extension)

---
In reply to: [725698360](https://github.com/microsoft/onnxruntime/pull/5771#issuecomment-725698360) [](ancestors = 725698360)

satyajandhyala commented:
I didn't change it. I however did rebase my branch before I pushed. on to the master to avoid any conflicts.

---
In reply to: [725696948](https://github.com/microsoft/onnxruntime/pull/5771#issuecomment-725696948) [](ancestors = 725696948)

skottmckay commented:
it's a diff in the PR so it must have been changed at some point. can you please revert to the commit id for onnx in master?

---
In reply to: [725724465](https://github.com/microsoft/onnxruntime/pull/5771#issuecomment-725724465) [](ancestors = 725724465,725696948)

skottmckay commented:
PR has onnxruntime/test/framework/testdata/mnist/model.ort

---
In reply to: [725723106](https://github.com/microsoft/onnxruntime/pull/5771#issuecomment-725723106) [](ancestors = 725723106,725698360)

satyajandhyala commented:
I was using mnist.onnx and mnist.ort to run an test I added, which later removed. I removed model.ort.

---
In reply to: [725727167](https://github.com/microsoft/onnxruntime/pull/5771#issuecomment-725727167) [](ancestors = 725727167,725723106,725698360)

satyajandhyala commented:
> 

The input_names (originally inputNames) and output_names (originally outputNames) are populated in init function. using resize instead of reserve is consistent with, the code populating these vectors by using indexing/assigning rather than push_back. If this were to be reserve instead of resize the indexing will fail. I think it is somewhat better to use resize and populate using push_back but it requires changing multiple places.

---
In reply to: [725704640](https://github.com/microsoft/onnxruntime/pull/5771#issuecomment-725704640) [](ancestors = 725704640)

---
Refers to: onnxruntime/test/fuzzing/src/OnnxPrediction.cpp:59 in 221eebd. [](commit_id = 221eebdae305cc046020727d4649b44d6f556cc1, deletion_comment = False)

skottmckay reviewed (approved):
:shipit:

skottmckay commented on onnxruntime/test/framework/ort_model_only_test.cc:
```diff
@@ -236,7 +236,39 @@ static void DumpOrtModelAsJson(const std::string& model_uri) {
 }
 */
+/* The full build was causing the following error because the graph node array has some empyt (blan) node at some indices for certain ORT designs
+       onnx runtime exception : Satisfied, but should not be : node == nullptr
```
Would be better to document why it happened vs. the effect of the issue. Bug was due to loading an ORT format model in a full build, allowing optimizers to run, but trying to use the saved kernel information. Because the optimizers removed nodes you get a mismatch.

skottmckay reviewed (approved):
:shipit:

gwang-msft commented on tools/ci_build/github/azure-pipelines/win-ci-fuzz-testing.yml:
```diff
@@ -0,0 +1,157 @@
+    buildArch: x64
+    setVcvars: true
+    ALLOW_RELEASED_ONNX_OPSET_ONLY: '0'
+  timeoutInMinutes: 2160
```
>2160 [](start = 20, length = 4)

Is this a bit too long for timeout?

satyajandhyala replied:
Earlier the build pipeline timed-out. I wanted find out how long it will run. I set a higher limit. I was not meant to checkin.


---
In reply to: [524839479](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r524839479) [](ancestors = 524839479)

gwang-msft commented on onnxruntime/core/graph/graph.cc:
```diff
@@ -3630,7 +3630,11 @@ common::Status Graph::LoadFromOrtFormat(const onnxruntime::experimental::fbs::Gr
+      if (node->Index() < fbs_graph.max_node_index()) {
+        nodes_[node->Index()] = std::move(node);
+      } else {
+        throw(std::out_of_range("Node index is out of range"));
```
>throw [](start = 8, length = 5)

ORT_THROW 

Or use ORT_ENFORCE on node->Index() < fbs_graph.max_node_index()

gwang-msft commented on onnxruntime/core/graph/graph.cc:
```diff
@@ -3640,7 +3644,11 @@ common::Status Graph::LoadFromOrtFormat(const onnxruntime::experimental::fbs::Gr
+      if (fbs_node_edge->node_index() < static_cast<size_t>(num_of_nodes_)) {
+        ORT_RETURN_IF_ERROR(nodes_[fbs_node_edge->node_index()]->LoadEdgesFromOrtFormat(*fbs_node_edge, *this));
+      } else {
+        throw(std::out_of_range("Node index is out of range"));
```
>throw [](start = 8, length = 5)

Same here

gwang-msft commented on onnxruntime/test/framework/ort_model_only_test.cc:
```diff
@@ -237,6 +237,41 @@ static void DumpOrtModelAsJson(const std::string& model_uri) {
   std::ofstream(model_uri + ".json") << json;
 }
 */
+/* The full build was causing the following error because the graph node array has some empyt (blan) node at some indices for certain ORT designs
```
>blan [](start = 95, length = 4)

blank?

gwang-msft commented on onnxruntime/core/session/inference_session.cc:
```diff
@@ -1131,30 +1131,34 @@ common::Status InferenceSession::Initialize() {
-    // Update temporary copies of metadata, input- and output definitions to the same state as the resolved graph
-    ORT_RETURN_IF_ERROR_SESSIONID_(SaveModelMetadata(*model_));
+    if (!loading_ort_format) {
+      // add predefined transformers
```
DO you want to add some comments here about why we not run optimizer when load ort format

skottmckay replied:
Should be covered by my PR (that's now checked in)

---
In reply to: [524842508](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r524842508) [](ancestors = 524842508)

satyajandhyala replied:
To avoid unnecessary conflicts due to the comments, I am not adding any of my own comments.


---
In reply to: [524842508](https://github.com/microsoft/onnxruntime/pull/5771#discussion_r524842508) [](ancestors = 524842508)

gwang-msft commented on onnxruntime/test/fuzzing/src/OnnxPrediction.cpp:
```diff
@@ -2,144 +2,126 @@
+#include <vector>
+
+static const char* const kOrtSessionOptionsConfigLoadModelFormat = "session.load_model_format";
+static const char* const kOrtSessionOptionsConfigSaveModelFormat = "session.save_model_format";
```
Please include the onnxruntime_session_options_config_keys.h

skottmckay commented on onnxruntime/test/framework/ort_model_only_test.cc:
```diff
@@ -237,6 +237,41 @@ static void DumpOrtModelAsJson(const std::string& model_uri) {
+  };
+
+  RunOrtModel(test_info);
+}
```
You don't need to run the model to repro the error due you? Just need to load the ort format model as it breaks during initialize. If so, can delete all this code around model execution.

satyajandhyala replied:
I want to execute the inference also. It is safe to test.

skottmckay commented on onnxruntime/test/fuzzing/src/test.cpp:
```diff
@@ -265,143 +239,170 @@ int processCommandLine(int argc, char* argv[], runtimeOpt& opt)
+    } else {
+      std::wstring ort_model_file = model_file;
+      if (model_file.substr(model_file.find_last_of(L".") + 1) == L"onnx") {
+        ort_model_file = model_file + L".ort";
```
Can we make the extension checks case insensitive?

Also add comment saying that the code here is generating an ORT format model from the ONNX model as that won't be obvious unless you know a lot about ORT.

skottmckay commented on onnxruntime/test/fuzzing/src/test.cpp:
```diff
@@ -265,143 +239,170 @@ int processCommandLine(int argc, char* argv[], runtimeOpt& opt)
-  catch(...)
-  {
-    Logger::testLog<< L"Something Went very wrong: " << Logger::endl;
+  } catch (const Ort::Exception& ort_exception) {
```
Should this return a non-zero code if there are non-ORT exceptions? Wondering whether the CI would fail due to a fuzz testing error if we don't return that info.

satyajandhyala replied:
There is a "return 1" at line #408. Doesn't it return non-zero? what am I missing?

skottmckay replied:
Sorry - github/codeflow mismatch. was referring to the line 399 that has 'return 0' so any errors from the fuzz testing loop with the try/catch don't result in a non-zero return.

skottmckay commented on tools/ci_build/github/azure-pipelines/win-ci-fuzz-testing.yml:
```diff
@@ -0,0 +1,157 @@
+        BuildConfig: 'Debug'
+      release:
+        BuildConfig: 'RelWithDebInfo'
+  variables:
```
Do we need two builds for this or would one suffice?

skottmckay commented on onnxruntime/test/fuzzing/src/OnnxPrediction.cpp:
```diff
@@ -2,144 +2,125 @@
-  inputNames.resize(ptrSession->GetInputCount());
-  outputNames.resize(ptrSession->GetOutputCount());
+  // TODO Use reserve instead of resize
+  input_names.resize(ptr_session->GetInputCount());
```
nit: Can this TODO be to do the initialization of input_names and output_names in init() instead of doing resize here and in other ctors, and adding the actual values in init()?

skottmckay commented on onnxruntime/test/fuzzing/src/OnnxPrediction.cpp:
```diff
@@ -2,144 +2,125 @@
+  raw_model = std::shared_ptr<void>(alloc.Alloc(onnx_model.ByteSizeLong()), [this](void* ptr) {
+    this->GetAllocator().Free(ptr);
+  });
 
```
nit: Add TODO saying this doesn't need to use shared_ptr or alloc (can be done with 'new' or a std::vector based buffer.

skottmckay commented on tools/ci_build/github/azure-pipelines/win-ci-fuzz-testing.yml:
```diff
@@ -0,0 +1,157 @@
+    inputs:
+      scriptPath: '$(Build.SourcesDirectory)\tools\ci_build\build.py'
+      arguments: '--gen_doc --config $(BuildConfig) --build_dir $(Build.BinariesDirectory) --skip_submodule_sync --build_shared_lib --update --cmake_generator "Visual Studio 16 2019" --build_wheel --use_dnnl --use_winml --use_openmp --build_shared_lib --enable_onnx_tests --enable_wcos --build_java --build_nodejs --use_full_protobuf --fuzz_testing'
+      workingDirectory: '$(Build.BinariesDirectory)'
```
Can we reduce what we're building here to just what's needed so it's quicker. I would guess these are not necessary
`-use_dnnl --use_winml --use_openmp --enable_onnx_tests --enable_wcos --build_java --build_nodejs`

skottmckay commented on tools/ci_build/github/azure-pipelines/win-ci-fuzz-testing.yml:
```diff
@@ -0,0 +1,157 @@
+      configuration: '$(BuildConfig)'          
+      arguments: '--configuration $(BuildConfig) -p:Platform="Any CPU" -p:OnnxRuntimeBuildDirectory="$(Build.BinariesDirectory)" -p:OrtPackageId=$(OrtPackageId) --blame'
+      workingDirectory: '$(Build.SourcesDirectory)\csharp'
+
```
We don't need to build and test C# do we?

satyajandhyala replied:
I copied the job from the existing YAML file. In general because we are changing some important files and build with some unique options, it is safe to run these tests.

skottmckay commented on tools/ci_build/github/azure-pipelines/win-ci-fuzz-testing.yml:
```diff
@@ -0,0 +1,157 @@
+  - template: templates/component-governance-component-detection-steps.yml
+    parameters :
+      condition : 'succeeded'
+  - task: CmdLine@2
```
Guessing we don't need to run the full unit tests either or publish those test results as that's covered by other CIs

skottmckay commented on tools/ci_build/github/azure-pipelines/win-ci-fuzz-testing.yml:
```diff
@@ -0,0 +1,157 @@
+    displayName: 'Run fuzz testing '
+    inputs:
+      script: '$(Build.BinariesDirectory)\$(BuildConfig)\$(BuildConfig)\onnxruntime_security_fuzz.exe /t  /f "$(Build.BinariesDirectory)\$(BuildConfig)\$(BuildConfig)\testdata\mnist.onnx"  1 m'
+      workingDirectory: $(Build.BinariesDirectory)\$(BuildConfig)\$(BuildConfig)
```
Can we add some more onnx models? Just testing one is a great start but may not have enough coverage. e.g. mnist has no subgraphs so there are sections of code that won't be touched. Can be separate PR.
